### PR TITLE
Add `import-profile-has-href-attribute` constraint

### PR DESCRIPTION
**Pull request by Gabeblis**
# Committer Notes
## Description
Added import-profile-has-href-attribute constraint. This should be added as it is part of the effort write all of the constraints from the constraint tracker.

## Changes
- Added import-profile-has-href-attribute constraint to fedramp-external-constraints.xml.
- Added pass and fail yaml tests for the above constraint.
- Added `ssp-import-profile-has-href-attribute-INVALID.xml` file for the fail case of this constraint.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?
- ~If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated?~ This PR references valid existing docs in the site.

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 1]
```diff
@@ -141,6 +141,10 @@
                 <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-name-abbreviation-and-fedramp-unique-identifier"/>
                 <message>A FedRAMP SSP must have a FedRAMP system identifier.</message>
             </expect>
+            <expect id="import-profile-has-href-attribute" target="import-profile" test="exists(doc(resolve-uri(@href))//control/prop[@ns='https://fedramp.gov/ns/oscal' and @name='CORE' and @value='true'])" level="CRITICAL">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/3-working-with-oscal-files/#importing-the-fedramp-baseline"/>
+                <message>A FedRAMP SSP MUST have an import-profile element that has a reference to a valid catalog with resolved controls in it.</message>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:146]: Nice work, I would recommend a slight adjustment to the wording, re [FRR113](https://github.com/GSA/fedramp-automation/blob/feature-constraints-style-guide/src/validations/constraints/STYLE.md#frr113)

```suggestion
                <message>A FedRAMP SSP MUST import a profile or catalog of security controls to reference implemented requirements against those control(s).</message>
```

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
Other than the slight message recommendation, nice work. I can do a quick re-review.

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
Actually, I forgot an important thing regarding the issue goal and ACs: can we talk tomorrow about implementing with an additional constraint to check for one or more properties of a control specific to FedRAMP? We have the power to do so, and that makes the check even more effective.

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 1 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml]: Thank you for catching this 🤦.

**Comment by Gabeblis** [comment]
> Actually, I forgot an important thing regarding the issue goal and ACs: can we talk tomorrow about implementing with an additional constraint to check for one or more properties of a control specific to FedRAMP? We have the power to do so, and that makes the check even more effective.

Sounds good to me.

**Comment by aj-stein-gsa** [comment]
> > Actually, I forgot an important thing regarding the issue goal and ACs: can we talk tomorrow about implementing with an additional constraint to check for one or more properties of a control specific to FedRAMP? We have the power to do so, and that makes the check even more effective.
> 
> 
> 
> Sounds good to me. 

Separate issue maybe?

That question aside, should we consider workshopping this tomorrow if you're up for it?

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 2]
```diff
@@ -141,6 +141,10 @@
                 <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-name-abbreviation-and-fedramp-unique-identifier"/>
                 <message>A FedRAMP SSP must have a FedRAMP system identifier.</message>
             </expect>
+            <expect id="import-profile-has-href-attribute" target="import-profile" test="exists(doc(resolve-uri(@href))//control/prop[@ns='https://fedramp.gov/ns/oscal' and @name='CORE' and @value='true'])" level="CRITICAL">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/3-working-with-oscal-files/#importing-the-fedramp-baseline"/>
+                <message>A FedRAMP SSP MUST import a profile or catalog of security controls to reference implemented requirements against those control(s).</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:147]: Blocking: this work is excellent a I really appreciate the ingenuity, but per the discussion while pairing today, can we split out the href check and then add a separate critical check like so (not complete, but you get the idea).

```suggestion
            <expect id="import-profile-has-href-attribute" target="import-profile" test="doc-available(@href))" level="CRITICAL">
                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/3-working-with-oscal-files/#importing-the-fedramp-baseline"/>
                <message>A FedRAMP SSP MUST import a profile or catalog with a valid file or HTTP(S) address.
            </expect>
            <expect id="tbd" target="import-profile" test="count(doc(resolve-uri(@href))//control/prop[@ns='https://fedramp.gov/ns/oscal' and @name='response-point']) &gte; 2">
                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/3-working-with-oscal-files/#importing-the-fedramp-baseline"/>
                <message>A FedRAMP SSP MUST import a profile or catalog of security controls to reference implemented requirements against those control(s).</message>               
            </expect>
```

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
Excellent work, just wanted to reiterate the feedback from pairing this afternoon:

1. Let's split out the constraints and use `response-point` counting.
2. I cannot call this out in a PR, but can you point to the profile in `src`, not `dist` in a separate commit from 1 and in commit for 2 explain why in the messgae. It will be useful later.

**Comment by Rene2mt** [comment]
The `import` "href" can be specified as:

1. a URI reference to a profile 
2. a URI reference to a resolved profile catalog
3. a fragment that points to a profile in the `back-matter`
4. a fragment that points to a resolved profile catalog in the `back-matter`  

Can we confirm if the Metapth syntax `doc(resolve-uri(@href))` treats all four of the above the same?

**Comment by aj-stein-gsa** [comment]
> Can we confirm if the Metapth syntax `doc(resolve-uri(@href))` treats all four of the above the same?

Personally I would love if we could add tests for that, positive and negative. @Gabeblis, do you understand what @Rene2mt is hinting at and think you can handle it in positive tests at least (not sure negative test would help there 😆).

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 2 above)
Comment (kyhu65867) [reply src/validations/constraints/fedramp-external-constraints.xml:147]: For what it's worth I like this, and think it is better to have more detailed error messages especially in a critical case. ^^

**Comment by Gabeblis** [comment]
> > Can we confirm if the Metapth syntax `doc(resolve-uri(@href))` treats all four of the above the same?
> 
> Personally I would love if we could add tests for that, positive and negative. @Gabeblis, do you understand what @Rene2mt is hinting at and think you can handle it in positive tests at least (not sure negative test would help there 😆).

I do understand the concept of what he's asking, but I'm not sure 

> > Can we confirm if the Metapth syntax `doc(resolve-uri(@href))` treats all four of the above the same?
> 
> Personally I would love if we could add tests for that, positive and negative. @Gabeblis, do you understand what @Rene2mt is hinting at and think you can handle it in positive tests at least (not sure negative test would help there 😆).

I do understand the concept of what Rene mentioned, but I am not sure how a fragment would point to a profile in the back-matter. I might need some clarification on how that'd show up.

**Comment by aj-stein-gsa** [comment]
> > > Can we confirm if the Metapth syntax `doc(resolve-uri(@href))` treats all four of the above the same?
> > 
> > 
> > Personally I would love if we could add tests for that, positive and negative. @Gabeblis, do you understand what @Rene2mt is hinting at and think you can handle it in positive tests at least (not sure negative test would help there 😆).
> 
> I do understand the concept of what Rene mentioned, but I am not sure how a fragment would point to a profile in the back-matter. I might need some clarification on how that'd show up.

I am really sorry I did not respond sooner. I should have clarified so you can, you know, right those tests. 😆 

I thought the FedRAMP baselines "ready for consumption" use this UUID-based back-matter approach, but it would seem not anymore.

Import:

https://github.com/GSA/fedramp-automation/blob/fedramp1.1.0-oscal1.0.0/dist/content/baselines/rev4/xml/FedRAMP_rev4_LOW-baseline_profile.xml#L54

Back-matter reference:

https://github.com/GSA/fedramp-automation/blob/07444ef9a83de63342e4f3655adb95fe79ddab04/dist/content/baselines/rev4/xml/FedRAMP_rev4_LOW-baseline_profile.xml#L6226C7-L6226C61

It seems we still use this approach:

https://github.com/GSA/fedramp-automation/blob/9d4d487e8e5e103018afcd1307bbeb111ff88d3b/dist/content/rev5/baselines/xml/FedRAMP_rev5_LI-SaaS-baseline_profile.xml#L51

https://github.com/GSA/fedramp-automation/blob/9d4d487e8e5e103018afcd1307bbeb111ff88d3b/dist/content/rev5/baselines/xml/FedRAMP_rev5_LI-SaaS-baseline_profile.xml#L2958

Does that clarify? I can help you make the relevant tests around that, I do not think our test harness does this like the production baselines.

**Review by aj-stein-gsa — COMMENTED** [review]
I like what I've seen so far. I assume, re chat messages, what we should do with test infra changes and if that's minor or not.

**Comment by Gabeblis** [comment]
> I like what I've seen so far. I assume, re chat messages, what we should do with test infra changes and if that's minor or not.

It is a pretty minor change. It is just a matter of if this is how we want this constraint/similar constraints to look. There is the option to create separate yaml files for each test scenario (I implemented this option). We could also have multiple files defined in the content of a single yaml test file for the constraint.

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario1-PASS.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+test-case:
+  name: Positive Test for import-profile-has-valid-content
+  description: >-
+    This test case validates the behavior of constraint
+    import-profile-has-valid-content
```
Comment (aj-stein-gsa) [inline src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario1-PASS.yaml:5]: Minor feedback: if we are going to put a description, I would say it makes sense to describe the unique scenario since we are testing four presumably for a good reason.

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario2-PASS.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+test-case:
+  name: Positive Test for import-profile-has-valid-content
+  description: >-
+    This test case validates the behavior of constraint
```
Comment (aj-stein-gsa) [inline src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario2-PASS.yaml:4]: Ditto re description.

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario3-PASS.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+test-case:
+  name: Positive Test for import-profile-has-valid-content
+  description: >-
+    This test case validates the behavior of constraint
+    import-profile-has-valid-content
```
Comment (aj-stein-gsa) [inline src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario3-PASS.yaml:5]: Ditto re description.

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario4-PASS.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+test-case:
+  name: Positive Test for import-profile-has-valid-content
+  description: >-
+    This test case validates the behavior of constraint
+    import-profile-has-valid-content
```
Comment (aj-stein-gsa) [inline src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario4-PASS.yaml:5]: Ditto re description.

**Review by wandmagic — CHANGES_REQUESTED** [review]
I see we allow for multiple positive cases but not multiple negative cases, I think we should treat them both equally

**Comment by wandmagic** [comment]
Does it make sense to allow content in the yaml to be an array?

# Original format
test-case:
  name: Negative Test for user-type
  description: This test case validates the behavior of constraint user-type
  content: ../content/ssp-user-type-INVALID.xml
  expectations:
    - constraint-id: user-type
      result: fail

# Modified format with array of content
test-case:
  name: Negative Test for user-type
  description: This test case validates the behavior of constraint user-type
  content:
    - ../content/ssp-user-type-INVALID1.xml
    - ../content/ssp-user-type-INVALID2.xml
    - ../content/ssp-user-type-INVALID3.xml
    - ../content/valid-user-type.xml
  expectations:
    - constraint-id: user-type
      result: fail

**Comment by Gabeblis** [comment]
> Does it make sense to allow content in the yaml to be an array?
> 
> # Original format
> ```
> test-case:
>   name: Negative Test for user-type
>   description: This test case validates the behavior of constraint user-type
>   content: ../content/ssp-user-type-INVALID.xml
>   expectations:
>     - constraint-id: user-type
>       result: fail
> ```
> 
> # Modified format with array of content
> ```
> test-case:
>   name: Negative Test for user-type
>   description: This test case validates the behavior of constraint user-type
>   content:
>     - ../content/ssp-user-type-INVALID1.xml
>     - ../content/ssp-user-type-INVALID2.xml
>     - ../content/ssp-user-type-INVALID3.xml
>     - ../content/valid-user-type.xml
>   expectations:
>     - constraint-id: user-type
>       result: fail
> ```

I am in favor of this approach

**Comment by Gabeblis** [comment]
> I see we allow for multiple positive cases but not multiple negative cases, I think we should treat them both equally

Agreed. I just didn't want to go tweaking a whole bunch of the test infrastructure before we came up with a decision for how to proceed with this constraint. This would be a good parking lot item to discuss today

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario4-PASS.yaml (hunk 6 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario4-PASS.yaml]: resolved: 1c3e3d3

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario2-PASS.yaml (hunk 4 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario2-PASS.yaml]: resolved: 1c3e3d3

File: src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario1-PASS.yaml (hunk 3 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/import-profile-has-valid-content-scenario1-PASS.yaml]: Also, (you guessed it), resolved: 1c3e3d3

**Review by aj-stein-gsa — COMMENTED** [review]
I talked over what I saw during the review because I only addressed the questions around the unit test structure. This constraint structure and messaging is good, I am happy with it.

Is the checklist updated and this is no longer draft? If the checklist is updated and taken out of draft state, I shall approve myself. 😄

**Comment by Gabeblis** [comment]
> > I see we allow for multiple positive cases but not multiple negative cases, I think we should treat them both equally
> 
> Agreed. I just didn't want to go tweaking a whole bunch of the test infrastructure before we came up with a decision for how to proceed with this constraint. This would be a good parking lot item to discuss today

@wandmagic I added the functionality to allow content in the yaml to be an array here [0fd885d](https://github.com/GSA/fedramp-automation/pull/817/commits/0fd885d25a675a7003cd51d58808181768cfb8e0). Thank you for the suggestion. I'd love some feedback when you have a chance to take a look 😄

**Comment by Gabeblis** [comment]
> I talked over what I saw during the review because I only addressed the questions around the unit test structure. This constraint structure and messaging is good, I am happy with it.
> 
> Is the checklist updated and this is no longer draft? If the checklist is updated and taken out of draft state, I shall approve myself. 😄

Checklist has been updated, and this is no longer a draft. Thank you sir 🫡

**Review by aj-stein-gsa — APPROVED** [review]

**Review by aj-stein-gsa — APPROVED** [review]

**Comment by vmangat** [comment]
Import-profile   should not be referencing a 'profile resolved catalog'.   This would be a catalog object not a profile object and violates the OSCAL standard.

**Comment by aj-stein-gsa** [comment]
> Import-profile should not be referencing a 'profile resolved catalog'. This would be a catalog object not a profile object and violates the OSCAL standard.

It is a should (the word used is discouraged) not a must, as [the guidance says](https://pages.nist.gov/OSCAL-Reference/models/v1.1.2/system-security-plan/xml-reference/#/system-security-plan/import-profile), so it is not forbidden. You will see the code permits for both.
